### PR TITLE
cluster up: add proxy support

### PR DESCRIPTION
**Pull request by csrwng**
Adds arguments to cluster up to set http, https proxy and no_proxy

Fixes https://github.com/openshift/origin/issues/11323

### CODE REVIEW

**Comment by csrwng** [comment]
@bparees fyi

File: pkg/bootstrap/docker/up.go [hunk 1]
```diff
@@ -149,6 +149,9 @@ func NewCmdUp(name, fullName string, f *osclientcmd.Factory, out, errout io.Writ
 	cmd.Flags().StringArrayVarP(&config.Environment, "env", "e", config.Environment, "Specify a key-value pair for an environment variable to set on OpenShift container")
 	cmd.Flags().BoolVar(&config.ShouldInstallMetrics, "metrics", false, "If true, install metrics (experimental)")
 	cmd.Flags().BoolVar(&config.ShouldInstallLogging, "logging", false, "If true, install logging (experimental)")
+	cmd.Flags().StringVar(&config.HTTPProxy, "http-proxy", "", "HTTP proxy to use for master and builds")
+	cmd.Flags().StringVar(&config.HTTPProxy, "https-proxy", "", "HTTPS proxy to use for master and builds")
+	cmd.Flags().StringArrayVar(&config.NoProxy, "no-proxy", config.NoProxy, "List of hosts or subnets for which a proxy should not be used")
```
Comment (bparees) [inline pkg/bootstrap/docker/up.go]: is it a list, or do you pass the arg multiple times, one per value?

File: docs/man/man1/oc-cluster-up.1 [hunk 2]
```diff
@@ -74,6 +74,10 @@ A public hostname can also be specified for the server with the \-\-public\-host
     If true, install metrics (experimental)
 
 .PP
+\fB\-\-persistent\-volumes\-dir\fP="/var/lib/origin/openshift.local.pv"
+    Directory on host for OpenShift persistent volumes
+
+.PP
```
Comment (bparees) [inline docs/man/man1/oc-cluster-up.1]: ??

File: docs/man/man1/openshift-cli-cluster-up.1 [hunk 3]
```diff
@@ -74,6 +74,10 @@ A public hostname can also be specified for the server with the \-\-public\-host
     If true, install metrics (experimental)
 
 .PP
+\fB\-\-persistent\-volumes\-dir\fP="/var/lib/origin/openshift.local.pv"
+    Directory on host for OpenShift persistent volumes
+
+.PP
```
Comment (bparees) [inline docs/man/man1/openshift-cli-cluster-up.1]: looks like you need to rerun docs gen?

File: pkg/bootstrap/docker/up.go (hunk 1 above)
Comment (csrwng) [inline pkg/bootstrap/docker/up.go]: You can pass it either a comma-separated list or specify multiple times, one per value

Comment (bparees) [inline pkg/bootstrap/docker/up.go]: k

**Comment by bparees** [comment]
@csrwng other than the question about the generated docs, lgtm

**Comment by bparees** [comment]
@csrwng will need this backported into 3.4 once it merges.

**Comment by csrwng** [comment]
Now using a fixed IP for the docker registry (`172.30.1.1`). This should make it easy to configure NO_PROXY for Docker. (It won't take a CIDR). 

Items remaining:

- [ ] Populate a default NO_PROXY value based on current host and registry CIDR
- [ ] Check that the Docker proxy settings are reasonable on startup (and warn otherwise).

**Comment by csrwng** [comment]
@bparees should be ready for another review

**Comment by bparees** [comment]
lgtm

**Comment by csrwng** [comment]
Tested on windows, mac and linux
Note on 'Docker for Windows': setting the proxy server for Docker causes every container created to inherit the proxy settings, which means that for any local services to be accessible from other containers, the NO_PROXY list configured on the Docker daemon needs to include their name and the .cluster.local prefix.

**Comment by csrwng** [comment]
[merge]

**Pull request merged**